server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs

```diff
@@ -49,110 +55,129 @@ import           Hasura.Prelude
 import           Hasura.RQL.Types
 import           Hasura.SQL.Value
 
-newtype RemoteRelKey =
-  RemoteRelKey Int
-  deriving (Eq, Ord, Show, Hashable)
-
-data RemoteRelField =
-  RemoteRelField
-    { rrRemoteField   :: !RemoteField
-    , rrArguments     :: !ArgsMap
-    , rrSelSet        :: !SelSet
+-- | When parameterised by 'RRF_Tree' this is like a 'VQ.Field' but with a path
```

Comment:
Maybe this is nitpicky but I don't like the name `RemoteRelBranch` :)

Should we call this `RemoteRelAnnField` to keep some semblance that this is basically a Field?

Also do we really need this kind of parameterisation? Is this for type-safety or for readability?

Reply:
Yeah the parameterisation is important (or once again making them two separate types); the two forms really are different things; the absence of `rrSelSet` means it's no longer a tree, and so definitely not `Field`-like.

But I'm not too happy with the names either.

Reply:
I find it really difficult to figure out what a function is doing (or supposed to be doing) if the domain is larger than it should be; the type signature becomes meaningless or misleading, you can't `traceShow` inputs and expect to understand what's going on, etc. 

So it's for readability I guess. Operationally it's probably also not good to hold on to a bunch of unneeded data.